lib/Monads/trace/Trace_Monad_Equations.thy

```diff
@@ -306,6 +340,28 @@ lemma if_bind:
 lemma bind_liftE_distrib: "(liftE (A >>= (\<lambda>x. B x))) = (liftE A >>=E (\<lambda>x. liftE (\<lambda>s. B x s)))"
   by (clarsimp simp: liftE_def bindE_def lift_def bind_assoc)
 
+(*FIXME: the following lemmas were originally solved by monad_eq, which doesn't yet exist for the
+         trace monad due to traces making equality more complicated.*)
```

Comment:
I think that's perfectly acceptable for the trace monad (not using `monad_eq`). Did you gain an impression on whether it would be feasible to update `monad_eq` in a way that would make sense for the trace monad? It's not used that much in the main refinement proofs, so it wouldn't be big loss, but if the work sparked an insight, it'd be good to know.

Reply:
I think it might be doable but it would be complicated.

The main issue is that for the nondet monad, equality can be nicely separated into `in_monad` style set membership and equality of the failure flag, both of which are resolved by rules we want anyway.

However, the trace monad version of `in_monad` only handles completed monad results, it doesn't say anything about the trace that leads to it or about `Failed` or `Incomplete` results. My guess is that with lemmas about those as well then we could update `monad_eq` to work for the trace monad, but there's a good chance that that would end up being more effort than it saves.

Reply:
Ok, we should definitely defer until we know we want this then, esp since `monad_eq` is not used that often.